### PR TITLE
Mimic constraint feature using bullet-featherstone

### PR DESCRIPTION
**Pull request by scpeters**
# 🎉 New feature

Targeted at #431 

## Summary

This adds support for `SetMimicConstraintFeature` in the `bullet-featherstone` plugin using `btMultiBodyGearConstraint`. The common test should be updated to use `bullet-featherstone`, and some error messages are missing.

## Test it

Building with https://github.com/gazebosim/gz-sim/pull/1838, load the gz-sim `examples/worlds/mimic_pendulum_world.sdf` and observe 3 pairs of pendulums with different lengths. The center pair of pendulums are uncoupled and oscillate at different frequencies. The left and right pairs have `<mimic/>` tags with `multiplier == 1` with different directionality of follower and leader and in both cases the long and short pendula should oscillate at the same frequency, and the frequency should be in between that of the uncoupled long and short pendula.

## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Comment by scpeters** [comment]
here's how the new test with fast and slow pendulums looks in gz-sim:

https://github.com/gazebosim/gz-physics/assets/3650755/59280958-1387-4a1f-a4c8-66df94f3c238

**Comment by scpeters** [comment]
this is ready for review. I think we should consider disabling `SetMimicConstraintFeature` in the `dartsim` plugin until it can pass the fast/slow pendulums test.

**Comment by scpeters** [comment]
I've retargeted this at `gz-physics6` and redacted the `dartsim` implementation since it is not yet working. I suggest we close https://github.com/gazebosim/gz-physics/pull/431 but leave the branch in place as a reference until https://github.com/dartsim/dart/issues/1756 is addressed

**Comment by mjcarroll** [comment]
Window trace:

```
 	gz-physics6-bullet-featherstone-plugin.dll!btCollisionWorld::~btCollisionWorld(void)	C++
>	gz-physics6-bullet-featherstone-plugin.dll!btMultiBodyDynamicsWorld::`vector deleting destructor'(unsigned int)	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::default_delete<btMultiBodyDynamicsWorld>::operator()(btMultiBodyDynamicsWorld *) Line 3167	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::unique_ptr<btMultiBodyDynamicsWorld,std::default_delete<btMultiBodyDynamicsWorld>>::{dtor}() Line 3278	C++
 	gz-physics6-bullet-featherstone-plugin.dll!gz::physics::bullet_featherstone::WorldInfo::~WorldInfo()	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::_Ref_count_base::_Decref() Line 1171	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::_Ptr_base<gz::physics::bullet_featherstone::WorldInfo>::_Decref() Line 1396	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>::{dtor}() Line 1680	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::_Default_allocator_traits<std::allocator<std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>>>::destroy(std::allocator<std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>> &) Line 686	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>::_Freenode(std::allocator<std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>> &) Line 318	C++
 	gz-physics6-bullet-featherstone-plugin.dll!std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>::_Free_non_head<std::allocator<std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>>>(std::allocator<std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *>> & _Head, std::_List_node<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,void *> *) Line 330	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::list<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,std::allocator<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>>>::_Tidy() Line 1498	C++
 	[Inline Frame] gz-physics6-bullet-featherstone-plugin.dll!std::list<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>,std::allocator<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>>>::{dtor}() Line 1043	C++
 	gz-physics6-bullet-featherstone-plugin.dll!std::_Hash<std::_Umap_traits<unsigned __int64,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>,std::_Uhash_compare<unsigned __int64,std::hash<unsigned __int64>,std::equal_to<unsigned __int64>>,std::allocator<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>>,0>>::~_Hash<std::_Umap_traits<unsigned __int64,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>,std::_Uhash_compare<unsigned __int64,std::hash<unsigned __int64>,std::equal_to<unsigned __int64>>,std::allocator<std::pair<unsigned __int64 const ,std::shared_ptr<gz::physics::bullet_featherstone::WorldInfo>>>,0>>()	C++
 	gz-physics6-bullet-featherstone-plugin.dll!gz::physics::bullet_featherstone::Plugin::`scalar deleting destructor'(unsigned int)	C++
 	[Inline Frame] gz-plugin2.dll!std::_Func_class<void,void *>::operator()(void * <_Args_0>) Line 861	C++
 	[Inline Frame] gz-plugin2.dll!gz::plugin::PluginWithDlHandle::{dtor}() Line 64	C++
 	[Inline Frame] gz-plugin2.dll!std::_Destroy_in_place(gz::plugin::PluginWithDlHandle &) Line 298	C++
 	gz-plugin2.dll!std::_Ref_count_obj2<gz::plugin::PluginWithDlHandle>::_Destroy() Line 2110	C++
 	[Inline Frame] gz-plugin2.dll!std::_Ref_count_base::_Decref() Line 1171	C++
 	[Inline Frame] gz-plugin2.dll!std::_Ptr_base<void>::_Decref() Line 1396	C++
 	[Inline Frame] gz-plugin2.dll!std::shared_ptr<void>::{dtor}() Line 1680	C++
 	[Inline Frame] gz-plugin2.dll!std::default_delete<gz::plugin::Plugin::Implementation>::operator()(gz::plugin::Plugin::Implementation *) Line 3167	C++
 	gz-plugin2.dll!std::unique_ptr<gz::plugin::Plugin::Implementation,std::default_delete<gz::plugin::Plugin::Implementation>>::~unique_ptr<gz::plugin::Plugin::Implementation,std::default_delete<gz::plugin::Plugin::Implementation>>() Line 3278	C++
 	gz-plugin2-loader.dll!gz::plugin::Plugin::`scalar deleting destructor'(unsigned int)	C++
 	[Inline Frame] COMMON_TEST_joint_mimic_features.exe!std::default_delete<gz::plugin::Plugin>::operator()(gz::plugin::Plugin *) Line 3167	C++
 	[Inline Frame] COMMON_TEST_joint_mimic_features.exe!std::unique_ptr<gz::plugin::Plugin,std::default_delete<gz::plugin::Plugin>>::{dtor}() Line 3278	C++
 	[Inline Frame] COMMON_TEST_joint_mimic_features.exe!gz::plugin::TemplatePluginPtr<gz::plugin::Plugin>::{dtor}() Line 57	C++
 	COMMON_TEST_joint_mimic_features.exe!JointMimicFeatureTest_PendulumsFastSlowMimicTest_Test::TestBody() Line 358	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2584	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::Test,void>(testing::Test * object, void(testing::Test::*)() method, const char * location) Line 2635	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::Test::Run() Line 2681	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::TestInfo::Run() Line 2856	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::TestSuite::Run() Line 3014	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::internal::UnitTestImpl::RunAllTests() Line 5871	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)()) Line 2584	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl,bool>(testing::internal::UnitTestImpl * object, bool(testing::internal::UnitTestImpl::*)()) Line 2635	C++
 	COMMON_TEST_joint_mimic_features.exe!testing::UnitTest::Run() Line 5444	C++
 	[External Code]	
```

**Comment by scpeters** [comment]
retargeted to harmonic

**Comment by scpeters** [comment]
this doesn't actually depend on `sdformat` for the mimic functionality, since it's not part of the `sdf` component, so I didn't have to rerun CI after merging https://github.com/gazebosim/sdformat/pull/1166

**Review by azeey — COMMENTED** [review]
Looks great! Mostly minor comments. My main concern is with the API taking strings instead of `Joint` objects and DOF values.

**Comment by scpeters** [comment]
> Looks great! Mostly minor comments. My main concern is with the API taking strings instead of `Joint` objects and DOF values.

thanks for the feedback; I've updated the API locally and am going to update gz-sim before pushing

File: test/common_test/worlds/mimic_universal_world.sdf [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    </physics>
+
+    <plugin
+      filename="gz-sim-physics-system"
+      name="gz::sim::systems::Physics">
+    </plugin>
+    <plugin
+      filename="gz-sim-scene-broadcaster-system"
+      name="gz::sim::systems::SceneBroadcaster">
+    </plugin>
```
[the remark this answers is not in the document]
Comment (scpeters) [reply test/common_test/worlds/mimic_universal_world.sdf]: https://github.com/gazebosim/gz-physics/pull/517/commits/d9c6480cc6b83acfa9668c027518574368374be4

File: test/common_test/worlds/mimic_prismatic_world.sdf [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+    </physics>
+
+    <plugin
+      filename="gz-sim-physics-system"
+      name="gz::sim::systems::Physics">
+    </plugin>
+    <plugin
+      filename="gz-sim-scene-broadcaster-system"
+      name="gz::sim::systems::SceneBroadcaster">
+    </plugin>
```
[the remark this answers is not in the document]
Comment (scpeters) [reply test/common_test/worlds/mimic_prismatic_world.sdf]: https://github.com/gazebosim/gz-physics/pull/517/commits/d9c6480cc6b83acfa9668c027518574368374be4

File: bullet-featherstone/src/JointFeatures.hh [hunk 3]
```diff
@@ -141,6 +143,16 @@ class JointFeatures :
   // ----- Transmitted wrench -----
   public: Wrench3d GetJointTransmittedWrenchInJointFrame(
       const Identity &_id) const override;
+
+  // ----- Mimic joint constraint -----
+  public: void SetJointMimicConstraint(
+      const Identity &_id,
+      const std::size_t _dof,
+      const std::string &_joint,
+      const std::string &_axis,
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.hh]: I've updated the API to pass a `BaseJointPtr` in https://github.com/gazebosim/gz-physics/pull/517/commits/d28ac5f35f5654d2c621733371311327ec9bab64 and updated gz-sim to use the new API in https://github.com/gazebosim/gz-sim/pull/1838/commits/e943e319bf0e925778a3d9bc0f1474bb95034b70

File: include/gz/physics/Joint.hh [hunk 4]
```diff
@@ -584,6 +586,52 @@ namespace gz
...
+      ///   The desired generalized coordinate corresponding to the follower
+      ///   axis within this joint. Values start from 0 and stop before
+      ///   Joint::GetDegreesOfFreedom().
+      /// \param[in] _joint
+      ///   name of the joint to be mimicked, i.e. the leader joint.
+      /// \param[in] _multiplier
+      ///   The multiplier to be applied to position of leader joint.
+      /// \param[in] _offset
+      ///   The offset to be applied to position of leader joint after
+      ///   the multiplier is applied.
```
[the remark this answers is not in the document]
Comment (scpeters) [reply include/gz/physics/Joint.hh]: fixed documentation as part of the API change in https://github.com/gazebosim/gz-physics/commit/d28ac5f35f5654d2c621733371311327ec9bab64

File: bullet-featherstone/src/JointFeatures.cc [hunk 5]
```diff
@@ -385,6 +385,76 @@ Wrench3d JointFeatures::GetJointTransmittedWrenchInJointFrame(
...
+  const auto leaderJointIt = model->jointNameToEntityId.find(_joint);
+  if (leaderJointIt == model->jointNameToEntityId.end())
+  {
+    // print error message
+    return;
+  }
+  auto leaderJointId =
+      this->GenerateIdentity(leaderJointIt->second,
+                             this->joints.at(leaderJointIt->second));
+  auto leaderJoint = this->ReferenceInterface<JointInfo>(leaderJointId);
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.cc]: remove as part of API change in https://github.com/gazebosim/gz-physics/commit/d28ac5f35f5654d2c621733371311327ec9bab64

File: bullet-featherstone/src/JointFeatures.cc [hunk 6]
```diff
@@ -385,6 +385,76 @@ Wrench3d JointFeatures::GetJointTransmittedWrenchInJointFrame(
...
   return wrenchOut;
 }
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointMimicConstraint(
+    const Identity &_id,
+    const std::size_t /*_dof*/,  // bullet doesn't mimic multi-axis joints yet
+    const std::string &_joint,
+    const std::string &/*_axis*/,
+    const double _multiplier,
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.cc]: removed as part of API change in https://github.com/gazebosim/gz-physics/commit/d28ac5f35f5654d2c621733371311327ec9bab64

File: bullet-featherstone/src/JointFeatures.cc [hunk 7]
```diff
@@ -385,6 +385,76 @@ Wrench3d JointFeatures::GetJointTransmittedWrenchInJointFrame(
...
+    const double _offset,
+    const double _reference)
+{
+  auto followerJoint = this->ReferenceInterface<JointInfo>(_id);
+  auto followerChild = this->ReferenceInterface<LinkInfo>(
+      followerJoint->childLinkID);
+  if (!followerChild->indexInModel.has_value())
+  {
+    // print error message
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.cc]: nice catch, updated to do this in https://github.com/gazebosim/gz-physics/pull/517/commits/9611e0988037acb2d5db9fc74e9cd38141266a4a

File: bullet-featherstone/src/JointFeatures.cc [hunk 8]
```diff
@@ -385,6 +385,76 @@ Wrench3d JointFeatures::GetJointTransmittedWrenchInJointFrame(
...
+      model->body.get(),
+      *followerChild->indexInModel,
+      model->body.get(),
+      *leaderChild->indexInModel,
+      btVector3(0, 0, 0),
+      btVector3(0, 0, 0),
+      btMatrix3x3::getIdentity(),
+      btMatrix3x3::getIdentity());
+  followerJoint->gearConstraint->setGearRatio(btScalar(-_multiplier));
+  followerJoint->gearConstraint->setRelativePositionTarget(
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.cc:454]: comments added in https://github.com/gazebosim/gz-physics/pull/517/commits/db893e24c3a6be17e208a4546eaa0499d2e6c59d

File: bullet-featherstone/src/JointFeatures.cc [hunk 9]
```diff
@@ -385,6 +385,76 @@ Wrench3d JointFeatures::GetJointTransmittedWrenchInJointFrame(
     jointInfo->jointFeedback->m_reactionForces.getAngular());
   return wrenchOut;
 }
+
+/////////////////////////////////////////////////
+void JointFeatures::SetJointMimicConstraint(
+    const Identity &_id,
+    const std::size_t /*_dof*/,  // bullet doesn't mimic multi-axis joints yet
```
[the remark this answers is not in the document]
Comment (scpeters) [reply bullet-featherstone/src/JointFeatures.cc]: I changed the return type to `bool` and added `gzerr` messages if `dof` != 0 in https://github.com/gazebosim/gz-physics/pull/517/commits/7b86015cb0390d1911291e7d4e0ae03f5c69c6b9

**Comment by scpeters** [comment]
I believe I've responded to your review comments

File: bullet-featherstone/src/JointFeatures.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (azeey) [reply bullet-featherstone/src/JointFeatures.cc:454]: Looks good. Thanks!

**Review by azeey — APPROVED** [review]

**Comment by scpeters** [comment]
there are some windows test failures, but I'm going to merge, since they are hard to track down

**Pull request merged**